Mathlib/Data/Multiset/Order.lean

```diff
@@ -0,0 +1,513 @@
+        have a0M: a0 ∈ M := by
+          apply mul_mem_not_erase
+          · change M = Multiset.erase (a0 ::ₘ X) a
+            rw [mul_singleton_erase, add_comm] at this0
+            exact this0
+          · exact hyp
+        apply mem_erase_cons
+        · exact a0M
+      exact H2
+
+lemma mord_wf_1 [DecidableEq α] [Preorder α] (a : α) (M0 : Multiset α)
+    (H1 : ∀ b M , LT.lt b a → AccM_1 M → AccM_1 (b ::ₘ M))
+    (H2 : AccM_1 M0)
+    (H3 : ∀ M, MultisetRedLt M M0 → AccM_1 (a ::ₘ M)) :
+    AccM_1 (a ::ₘ M0) := by
+  constructor
+  intros N N_lt
+  change AccM_1 N
+  rcases (red_insert N_lt) with ⟨x, H, H0⟩
+  case h.intro.inr h =>
+    rcases h with ⟨H, H0⟩
+    rw [H]
+    clear H -- Needed to make simp_all below safe.
+    induction x using Multiset.induction with
+    | empty =>
+      simpa
+    | cons h =>
+      simp_all only [Multiset.mem_cons, or_true, implies_true, true_implies, forall_eq_or_imp,
+        Multiset.add_cons]
+  case h.intro.inl.intro =>
+    simp_all
+
+lemma mord_wf_2 [DecidableEq α] [Preorder α] (a : α)
+    (H : ∀ (b : α), ∀ M, LT.lt b a → AccM_1 M → AccM_1 (b ::ₘ M)) :
+    ∀ M, AccM_1 M → AccM_1 (a ::ₘ M) := by
+  unfold AccM_1
+  intros M H0
+  induction H0 with
+  | intro x wfH wfH2 =>
+    apply mord_wf_1
+    · simpa
+    · constructor
+      simpa
+    · simpa
+
+lemma mord_wf_3[DecidableEq α] [Preorder α] :
+    ∀ (a:α), Acc LT.lt a → ∀ M, AccM_1 M → AccM_1 (a ::ₘ M) := by
+  intro w w_a
+  induction w_a with
+  | intro x _ ih =>
+      intro M accM1
+      apply @mord_wf_2 α _ _ _ _ _ accM1
+      simp_all
+
+/- If all elements of a multiset `M` are accessible with `LT.lt`, then the multiset M is
+accessible given the `MultisetRedLt` relation. -/
+lemma mred_acc [DecidableEq α] [Preorder α] :
+    ∀ (M : Multiset α), (∀x, x ∈ M → Acc LT.lt x) → AccM_1 M  := by
+  intros M wf_el
+  induction M using Multiset.induction_on with
+  | empty =>
+    constructor
+    intro y y_lt
+    absurd y_lt
+    apply not_MultisetRedLt_0
+  | cons _ _ ih =>
+    apply mord_wf_3
+    · apply wf_el
+      simp_all
+    · apply ih
+      intros
+      apply wf_el
+      simp_all
+
+/- If `LT.lt` is well-founded, then `MultisetRedLt` is well-founded. -/
+lemma RedLt_wf [DecidableEq α] [Preorder α]
+    (wf_lt : WellFoundedLT α) : WellFounded (MultisetRedLt : Multiset α → Multiset α → Prop) := by
+  constructor
+  intros a
+  apply mred_acc
+  intros x _
+  apply wf_lt.induction x
+  intros y h
+  apply Acc.intro y
+  assumption
+
+/- If `MultisetRedLt` is well-founded, then its transitive closure `MultisetLt` is also
+well-founded. -/
+lemma Lt_wf [DecidableEq α] [LT α]
+    (h : WellFounded (MultisetRedLt : Multiset α → Multiset α → Prop)) :
+    WellFounded (MultisetLt : Multiset α → Multiset α → Prop) := by
+  unfold MultisetLt
+  apply WellFounded.transGen
+  assumption
+
+lemma mem_leq_diff [DecidableEq α] : ∀ (M N: Multiset α), N ≤ M → M = M - N + N := by
```

Comment:
This and following need to move to an earlier file. 

Indeed, it may be best to make a separate PR with the basic Multiset material, so it can be reviewed properly separately.